Handle -1 in jagged layout NT view ops

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115842
* #115192
* #115191
* __->__ #115843
* #115636

Allows for inheriting the ragged and batch dims via -1:
```python
nt.view(-1, -1, D)
nt.expand(B, -1, D)
```

jbschlosser committed:
Handle -1 in jagged layout NT view ops

[ghstack-poisoned]

jbschlosser committed:
Update on "Handle -1 in jagged layout NT view ops"


TODO: tests!

[ghstack-poisoned]

jbschlosser committed:
Update on "Handle -1 in jagged layout NT view ops"


TODO: tests!

[ghstack-poisoned]

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -2033,8 +2033,8 @@ def test_transpose_inference_mode_interaction(self, device, dtype):
-    def test_view(self, device, dtype):
-        nt = random_nt(device, dtype, 4, (4, 4))
+    def test_view(self, device, dtype, layout):
+        nt = random_nt(device, dtype, 4, (4, 4), layout=layout)
```
I think you forgot to do `@parametrize` :P

jbschlosser replied:
yep this is completely wrong, I started do it and bailed lol

jbschlosser committed:
Update on "Handle -1 in jagged layout NT view ops"


Allows for inheriting the ragged and batch dims via -1:
```python
nt.view(-1, -1, D)
nt.expand(B, -1, D)
```

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
